### PR TITLE
Fix use of hyphen  in module name

### PR DESCRIPTION
**Pull request by hbcarlos**
As discussed in #9421 and #9576 there is an issue on the `federated_labextension.py` script.

There is already an opened PR #9577 for a temporary solution but this is meant to be a more robust solution.

## Code changes
When trying to get the extension metadata, if the argument module name is not a valid lab extension now the script uses `find_packages()` to look into all the modules in the package and returns the first one with valid lab extension metadata.

## User-facing changes
No user changes.

## Backwards-incompatible changes
No incompatibility.

### CODE REVIEW

**Comment by hbcarlos** [comment]
ping @jtpio @martinRenou

File: jupyterlab/federated_labextensions.py [hunk 1]
```diff
@@ -359,40 +359,49 @@ def _get_labextension_metadata(module):
...
-            m = importlib.import_module(package)
-
-    if not hasattr(m, '_jupyter_labextension_paths'):
-        raise KeyError('The Python module {} is not a valid labextension, '
-                       'it is missing the `_jupyter_labextension_paths()` method.'.format(module))
-    labexts = m._jupyter_labextension_paths()
-    return m, labexts
+    # Try getting the package name from setup.py
+    try:
+        package = subprocess.check_output([sys.executable, 'setup.py', '--name'], cwd=mod_path).decode('utf8').strip()
```
Comment (martinRenou) [inline jupyterlab/federated_labextensions.py:381]: We should maybe not even do this anymore. @maartenbreddels found that this was causing issues when the setup.py does some logging.

File: jupyterlab/federated_labextensions.py [hunk 2]
```diff
@@ -359,40 +359,49 @@ def _get_labextension_metadata(module):
...
+    except subprocess.CalledProcessError:
+        raise FileNotFoundError('The Python package `{}` is not a valid package, '
+                'it is missing the `setup.py` file.'.format(module))
+
+    # Make sure the package is installed
+    import pkg_resources
+    try:
+        dist = pkg_resources.get_distribution(package)
+    except pkg_resources.DistributionNotFound:
+        subprocess.check_call([sys.executable, '-m', 'pip', 'install', '-e', mod_path])
```
Comment (martinRenou) [inline jupyterlab/federated_labextensions.py:391]: Do we want to install the package?

File: jupyterlab/federated_labextensions.py [hunk 3]
```diff
@@ -359,40 +359,49 @@ def _get_labextension_metadata(module):
...
+
+    # Looking for modules in the package
+    from setuptools import find_packages
+    packages = find_packages(mod_path)
+
+    # Looking for the labextension metadata
+    for package in packages :
+        m = importlib.import_module(package)
+        if hasattr(m, '_jupyter_labextension_paths') :
+            return m, m._jupyter_labextension_paths()
```
Comment (martinRenou) [inline jupyterlab/federated_labextensions.py:402]: 👍🏼

File: jupyterlab/federated_labextensions.py (hunk 2 above)
Comment (hbcarlos) [reply jupyterlab/federated_labextensions.py:391]: Not sure, personally I install the python package before installing the extension but maybe someone is using `jupyter-labextensions develop .` to install everything.

Comment (jtpio) [reply jupyterlab/federated_labextensions.py:391]: Installing the package is the default behavior of the `develop` command at the moment.

**Comment by jtpio** [comment]
Thinking out loud. Maybe it will be possible to construct some test cases by generating extensions with hyphens and underscores in the names, using the cookiecutter.

**Comment by hbcarlos** [comment]
I'm working on some test extensions.

**Comment by jtpio** [comment]
> I'm working on some test extensions.

Thanks!

> Thinking out loud. Maybe it will be possible to construct some test cases by generating extensions with hyphens and underscores in the names, using the cookiecutter.

Actually it might also be simpler to stick to mock packages that are similar to the existing ones (and minimal)

**Review by hbcarlos — COMMENTED** [review]
Since this code is getting complicated, I propose to limit the options of users to modules with `__init__.py` so we can find the module with `setuptools.find_package()` or making the user indicating the module path when executing the script `jupyter labextensions develop path/to/module`.

File: jupyterlab/federated_labextensions.py [hunk 4]
```diff
@@ -390,6 +390,11 @@ def _get_labextension_metadata(module):
     except pkg_resources.DistributionNotFound:
         subprocess.check_call([sys.executable, '-m', 'pip', 'install', '-e', mod_path])
         sys.path.insert(0, mod_path)
+    
+    # Importing module with the same name as package
+    m = importlib.import_module(package)
+    if hasattr(m, '_jupyter_labextension_paths') :
+        return m, m._jupyter_labextension_paths()
```
[the remark this answers is not in the document]
Comment (martinRenou) [reply jupyterlab/federated_labextensions.py:397]: We could maybe find a way to get the name from the setup.py without running it

File: jupyterlab/tests/mock_packages/test-hyphens-underscore/pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,3 @@
+[build-system]
+requires = ["jupyter_packaging~=0.7.9", "jupyterlab>=3.0.0,==3.*", "setuptools>=40.8.0", "wheel"]
```
Comment (jtpio) [inline jupyterlab/tests/mock_packages/test-hyphens-underscore/pyproject.toml:2]: We might also want to use:

```suggestion
requires = ["jupyter_packaging~=0.7.9", "jupyterlab~=3.0", "setuptools>=40.8.0", "wheel"]
```

**Comment by jtpio** [comment]
Thanks Carlos! 

I pushed a couple more changes and rebased on the latest master, and CI seems to be passing now.

File: scripts/ci_script.sh [hunk 6]
```diff
@@ -190,6 +190,11 @@ if [[ $GROUP == usage ]]; then
     jupyter labextension develop extension --debug
     jupyter labextension build extension
 
+    # Test develop script with hyphens and underscores in the module name
+    jupyter labextension develop test-hyphens --overwrite --debug
```
Comment (jtpio) [inline scripts/ci_script.sh:194]: We might want to double check whether the `--overwrite` flag is needed here.

**Review by blink1073 — APPROVED** [review]
Thanks!

**Comment by blink1073** [comment]
@meeseeksdev please backport to 3.0.x

**Pull request merged**